tests/test_add_model.py

```diff
@@ -0,0 +1,423 @@
+                                                            'type': 'string'}},
+                              'required': ['arg1', 'arg2']},
+                                ..........
+    """
+    _mock_registry = {"models": {}}
+    mock_registry.__getitem__.side_effect = _mock_registry.__getitem__
+    mock_registry.__setitem__.side_effect = _mock_registry.__setitem__
+
+    swagger.add_model(test_input)
+
+    assert test_input.__name__ in _mock_registry["models"]
+    assert "description" in _mock_registry["models"][test_input.__name__]
+    assert "notes" in _mock_registry["models"][test_input.__name__]
+
+    if "resource_fields" in dir(test_input):
+        if hasattr(test_input, "required"):
+            assert "required" in _mock_registry["models"][test_input.__name__]
+    else:
+        assert "required" in _mock_registry["models"][test_input.__name__]
+
+    assert "properties" in _mock_registry["models"][test_input.__name__]
+
+
+# Setup test - ensure `_parse_doc(..)` is called without issues
+test_fixtures_model_inputs = [
+    MockBasicObject,
+    TodoItem,
+    ModelWithResourceFieldsNoRequired,
+    TodoItemWithResourceFields,
+]
+
+
+@patch("flask_restful_swagger.swagger.registry")
+@patch("flask_restful_swagger.swagger._parse_doc")
+@pytest.mark.parametrize("input_model", test_fixtures_model_inputs)
+def test_add_model_get_docs(mock_parse_doc, mock_registry, input_model):
+    _mock_registry = {"models": {}}
+    mock_registry.__getitem__.side_effect = _mock_registry.__getitem__
+    mock_registry.__setitem__.side_effect = _mock_registry.__setitem__
+    mock_parse_doc.return_value = (None, None)
+    swagger.add_model(input_model)
+
+    mock_parse_doc.assert_called_once_with(input_model)
+
+
+# Setup test for - with resource fields, no init, without swagger metadata.
+mock_with_resource_fields_with_required = MagicMock(
+    spec=ModelWithResourceFieldsWithRequired
+)
+mock_with_resource_fields_with_required.__name__ = "mock_name_attribute"
+mock_with_resource_fields_with_required.resource_fields.items = Mock(
+    return_value=dict(a=1, b=2).items()
+)
+
+
+test_fixtures_model_inputs = [
+    mock_with_resource_fields_with_required,
+]
+
+
+@patch("flask_restful_swagger.swagger.registry")
+@patch("flask_restful_swagger.swagger.deduce_swagger_type")
+@patch("flask_restful_swagger.swagger._Nested", spec=swagger._Nested)
+@patch("flask_restful_swagger.swagger.isinstance")
+@patch("flask_restful_swagger.swagger.hasattr")
+@patch("flask_restful_swagger.swagger.dir")
+@patch("flask_restful_swagger.swagger._parse_doc")
+@pytest.mark.parametrize("mock_model_class", test_fixtures_model_inputs)
+def test_add_model_with_resource_fields_without_swagger_metadata(
+    mock_parse_doc,
+    mock_dir,
+    mock_hasattr,
+    mock_isinstance,
+    mock_nested,
+    mock_deduce_swagger_type,
+    mock_registry,
+    mock_model_class,
+):
+    _mock_registry = {"models": {}}
```

Comment:
You might be able to put this setup material into a function, and avoid copying and pasting the whole block between tests.

Reply:
Refactored the data structure setup into seperate file `tests/fixtures_add_model.py`.  Using context managers to centralise the setup.